### PR TITLE
WIP: Kruskal's Algorithm to find minimum spanning tree

### PR DESCRIPTION
**Pull request by athulappadan**
Please give suggestions to improve the implementation.

### CODE REVIEW

File: src/spanningtrees/kruskal.jl [hunk 1]
```diff
@@ -0,0 +1,55 @@
+immutable KruskalHeapEntry{T}
+    edge::Pair{Int, Int}
```
Comment (sbromberger) [inline src/spanningtrees/kruskal.jl]: use `LightGraphs.Edge` here.

File: src/spanningtrees/kruskal.jl [hunk 2]
```diff
@@ -0,0 +1,55 @@
+immutable KruskalHeapEntry{T}
```
Comment (sbromberger) [inline src/spanningtrees/kruskal.jl]: `T <: Real`?

File: src/spanningtrees/kruskal.jl [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+function check_connected(nodes, p, q)
+
+    return nodes[p] == nodes[q]
+end
+
+"""Kruskal's algorithm
+to find minimum spanning tree for undirected graphs
+"""
+function kruskal_mst{T}(
```
Comment (sbromberger) [inline src/spanningtrees/kruskal.jl]: `T <: Real`?

File: src/spanningtrees/kruskal.jl [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+end
+
+function check_connected(nodes, p, q)
+
+    return nodes[p] == nodes[q]
+end
+
+"""Kruskal's algorithm
+to find minimum spanning tree for undirected graphs
+"""
```
Comment (sbromberger) [inline src/spanningtrees/kruskal.jl]: Needs a fuller docstring. What are the parameters, what gets returned, and where's the reference to the implementation?

File: src/spanningtrees/kruskal.jl [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+function kruskal_mst{T}(
+    g::SimpleGraph,
+    distmx::AbstractArray{T, 2}
+)
+
+    edge_list = []
```
Comment (sbromberger) [inline src/spanningtrees/kruskal.jl]: `edge_list = Vector{Edge}()` ? Also, if you have an idea how big this is going to be, use `sizehint!`.

File: src/spanningtrees/kruskal.jl [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    mst = []
```
Comment (sbromberger) [inline src/spanningtrees/kruskal.jl]: What type is this? Right now you're creating a vector of `Any` which is not optimal. From below, it looks like this is also `Vector{Edge}()`.

File: src/spanningtrees/kruskal.jl [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    connected_nodes = Vector{Int}(1:nv(g))
+
+    for e in edges(g)
+        println(e)
```
Comment (sbromberger) [inline src/spanningtrees/kruskal.jl]: is this `println` leftover from debug?

File: src/spanningtrees/kruskal.jl [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+        if nodes[i] == pid
+            nodes[i] = qid
+        end
+    end
+end
+
+function check_connected(nodes, p, q)
+
+    return nodes[p] == nodes[q]
+end
```
Comment (sbromberger) [inline src/spanningtrees/kruskal.jl]: Do you need this function?

File: src/spanningtrees/kruskal.jl [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+"""
+function quick_find(nodes, p, q)
+    pid = nodes[p]
+    qid = nodes[q]
+    for i in 1:length(nodes)
+        if nodes[i] == pid
+            nodes[i] = qid
+        end
+    end
+end
```
Comment (sbromberger) [inline src/spanningtrees/kruskal.jl]: This mutates nodes, so should have a `!` at the end of the function name. Also, the docstring needs a reference to the quick-find algorithm, along with inputs and expected behavior.

File: src/spanningtrees/kruskal.jl [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+        w = dst(e.edge)
+
+        if !check_connected(connected_nodes, v, w)
+            quick_find(connected_nodes, v, w)
+            push!(mst, e)
+        end
+    end
+
+    return mst
+end
```
Comment (sbromberger) [inline src/spanningtrees/kruskal.jl]: Needs tests as well.

**Comment by athulappadan** [comment]
Thanks for the suggestions. Will push the changes soon..

**Comment by CarloLucibello** [comment]
you may want to also add the tests in #277

**Comment by athulappadan** [comment]
Can someone explain why tests failed, please..?

**Comment by sbromberger** [comment]
https://travis-ci.org/JuliaGraphs/LightGraphs.jl/jobs/161864309#L479-L486 because of
https://github.com/JuliaGraphs/LightGraphs.jl/pull/456/files#diff-fce720c43af3c52c862fd7451c7374b8R92

Get rid of the `.jl` in `runtests` and you should make some progress.

**Comment by athulappadan** [comment]
Ok. Why src/LightGraphs.jl shown as a conflicting file?

**Comment by CarloLucibello** [comment]
git fetch origin  # assuming that origin is the remote name of this repo
git rebase -i origin/master

**Comment by sbromberger** [comment]
or just `git pull; git merge master` and then `git mergetool` to resolve the conflicts.

**Comment by athulappadan** [comment]
I am not so experienced with git and I am finding it difficult to resolve this conflict. Is it okay if I open a new PR? Sorry for the inconvenience.

**Comment by CarloLucibello** [comment]
sure, much better

Il giorno dom 25 set 2016 alle ore 07:11 Athul George <
notifications@github.com> ha scritto:

**Comment by athulappadan** [comment]
See #465

**Pull request closed without merging**